platform/pulp/platform/models/content.py

```diff
@@ -0,0 +1,134 @@
+from pulp.platform.models.storage import StoragePath
+
+
+class Content(MasterModel):
+    """
+    A piece of managed content.
+
+    :cvar natural_key_fields: Tuple of natural fields.  Must be: models.Field.
+    :type natural_key_fields: tuple
+
+    Fields:
+
+    :cvar type: The content type.
+    :type type: models.TextField
+
+    Relations:
+
+    :cvar notes: Arbitrary information stored with the content.
+    :type notes: fields.GenericRelation
+    """
+    natural_key_fields = ()
+
+    type = models.TextField(blank=False, default=None)
+
+    notes = fields.GenericRelation(Notes)
+
+    def natural_key(self):
+        """
+        Get the model's natural key based on natural_key_fields.
+
+        :return: The natural key.
+        :rtype: tuple
+        """
+        return (getattr(self, f.name) for f in self.natural_key_fields)
+
+    def natural_key_digest(self):
+        """
+        Get the SHA-256 digest of the natural key.
+        The digest is updated with each field name followed by its value.
+        The field names are only necessary to preserve backward compatibility
+        with digests generated in pulp 2.
+
+        :return: The hex digest.
+        :rtype: str
+        """
+        h = hashlib.sha256()
+        for name in sorted(f.name for f in self.natural_key_fields):
+            h.update(name)
+            value = getattr(self, name)
+            if not isinstance(value, basestring):
+                h.update(str(value))
+            else:
+                h.update(value)
+        return h.hexdigest()
+
+
+class Artifact(Model):
+    """
+    A file associated with a piece of content.
+
+    Fields:
+
+    :cvar file: The stored file.
+    :type file: models.FileField
+
+    :cvar downloaded: The associated file has been successfully downloaded.
+    :type downloaded: BooleanField
+
+    :cvar published_path: The relative published path.
+    :type published_path: models.FileField
+
+    :cvar size: The size of the file in bytes.
+    :type size: models.IntegerField
+
+    :cvar md5: The MD5 checksum of the file.
+    :type md5: models.CharField
+
+    :cvar sha1: The SHA-1 checksum of the file.
+    :type sha1: models.CharField
+
+    :cvar sha224: The SHA-224 checksum of the file.
+    :type sha224: models.CharField
+
+    :cvar sha256: The SHA-256 checksum of the file.
+    :type sha256: models.CharField
+
+    :cvar sha384: The SHA-384 checksum of the file.
+    :type sha384: models.CharField
+
+    :cvar sha512: The SHA-512 checksum of the file.
+    :type sha512: models.CharField
+
+    Relations:
+
+    :cvar content: The associated content.
+    :type content: Content
+    """
+
+    # Note: The FileField does not support unique indexes and has
+    # other undesirable behavior and complexities.  Using a custom
+    # field should be investigated.
+
+    file = models.FileField(db_index=True, upload_to=StoragePath(), max_length=255)
+    downloaded = models.BooleanField(db_index=True, default=False)
```

Comment:
Does it make sense to have a field on this model, say ``cached``, to indicate it's been handled by the streamer? Currently the ``DeferredDownload`` collection is just the unit_id and unit_type_id. Then the periodic task that downloads the cached content just queries all artifacts with ``cached = True`` and gets to the catalog entries by way of the foreign key relationship between ``LazyCatalogEntry`` and ``Artifact``.